util/feed.go

```diff
@@ -0,0 +1,396 @@
+				// "unsubscribed" entry and decrease the number of activeSubs
+				// for future entries.
+				addEntry(unsubEntry, pub.id)
+				f.activeCount--
+			case event := <-f.events:
+				// A new event has been published to the feed, which will
+				// eventually be read by all active subscribers.
+				if f.activeCount == 0 {
+					break
+				}
+				addEntry(valueEntry, event)
+			case <-stopper.ShouldStop():
+				// The stopper has been activated and the feed should cease.
+				if f.activeCount != 0 {
+					// Add a 'closed' entry to the linked channel, which will
+					// cause consuming publishers to stop.
+					addEntry(closedEntry, nil)
+				}
+				close(f.closer)
+				return
+			}
+		}
+	}()
+	return f
+}
+
+// Publish publishes a event into the Feed, which will eventually be received by
+// all Subscribers to the feed. Publishing to a closed feed will panic.
+func (f *Feed) Publish(event interface{}) {
+	select {
+	case f.events <- event:
+		// Event was successfully published.
+	case <-f.closer:
+		panic("attempt to publish value to a closed Feed")
+	}
+}
+
+// Subscribe returns a Subscription object which can immediately recieve events
+// which were published to this feed. An event is an arbitrary interface.
+//
+// Events are read from the Subscription's Events channel. Subscribers cannot
+// block each other from receiving events, but should still attempt to consume
+// events in a timely fashion; the Feed will not release the memory for an
+// individual event until every Subscriber has seen it.
+//
+// Attempting to subscribe to a closed Feed will panic.
+func (f *Feed) Subscribe() *Subscription {
+	// events and closer are used to manage the interaction between a
+	// Subscription and its backing publisher.
+	events := make(chan interface{})
+	closer := make(chan struct{})
+	sub := &Subscription{
+		Events: events,
+		closer: closer,
+	}
+	p := &publisher{
+		events: events,
+		closer: closer,
+		feed:   f,
+	}
+	select {
+	case f.subscribe <- p:
+		// Subscription was successfully logged.
+	case <-f.closer:
+		panic("attempt to publish value to a closed Feed")
+	}
+	return sub
+}
+
+// A publisher reads values from a Feed and passes them to a Subscriber; each
+// Subscriber is backed by a publisher. The publisher correctly reads from the
+// linked channel provided by the Feed, ensuring that other publishers cannot be
+// blocked.
+type publisher struct {
+	id      int
+	current *feedEntry
+	events  chan<- interface{}
+	closer  <-chan struct{}
+	feed    *Feed
+}
+
+// process is a method which reads the linked channel provided by a Feed,
+// passing relevant events to a Subscription. It is
+// intended to be run as a goroutine.
+func (p *publisher) process() {
+	var (
+		// If the Subscriber has unsubscribed, the publisher must still
+		// read events from the Feed until the unsubscription event has
+		// been acknowledged by the Feed: this ensures that each
+		// feedEntry reaches its consumerCount count and is returned to
```

Comment:
It is not strictly necessary to return feedEntries to the pool; there is no memory leak if you never call pool.Put. Those entries will simply be garbage collected as usual, which is less cpu-efficient, but if unsubscribes are uncommon then it might be better to forego the complexity of draining and just let them be GC'd.